### PR TITLE
List users and patients

### PR DESCRIPTION
**Pull request by arkadiuszbachorski**
# List users and patients

## :recycle: Current situation & Problem
Users and patients need to be listed. 

## :gear: Release Notes 
* Add features to DataTable
* Create "Users" list page
* Create "Patients list" page

## :white_check_mark: Testing
Patients and users views will be covered with E2E tests later, hence drop in code coverage

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -28,17 +28,22 @@
   },
   "dependencies": {
     "@hookform/resolvers": "^3.6.0",
+    "@nextui-org/pagination": "^2.0.33",
     "@stanfordbdhg/design-system": "file:./packages/design-system",
     "@t3-oss/env-nextjs": "^0.10.1",
+    "@tanstack/match-sorter-utils": "^8.15.1",
     "@tanstack/react-table": "^8.19.2",
     "firebase": "^10.12.2",
+    "firebase-admin": "^12.2.0",
```
Comment (PSchmiedmayer) [inline package.json:37]: TODO to ensure that we remove the admin SDK before we merge the PR; we should ensure that we don't use the admin SDK in the client-facing web dashboard.

File: modules/firebase/utils.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  organizations: 'organizations',
+}
+
+export const getCollectionRefs = (db: Firestore) => ({
+  users: () =>
+    collection(db, collectionNames.users) as CollectionReference<User>,
+  admins: () =>
+    collection(db, collectionNames.admins) as CollectionReference<Admin>,
+  clinicians: () =>
+    collection(db, collectionNames.admins) as CollectionReference<Clinician>,
```
Comment (PSchmiedmayer) [inline modules/firebase/utils.ts:53]: In connection to the changes that will soon be merged by @pauljohanneskraft we should ensure that we point this to `/clinicians`

https://github.com/StanfordBDHG/ENGAGE-HF-Firebase/pull/13/files

File: modules/firebase/utils.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+export const getAdminApp = () => {
+  try {
+    return getApp('adminApp')
+  } catch (error) {
+    return initializeApp({}, 'adminApp')
+  }
+}
+
+export interface Organization {
```
Comment (PSchmiedmayer) [inline modules/firebase/utils.ts:16]: We should setup an issue that we eventually re-use the types from the Firebase backend setup.

File: packages/design-system/src/components/CopyText/CopyText.tsx
Comment (PSchmiedmayer) [inline packages/design-system/src/components/CopyText/CopyText.tsx:1]: Would be cool to also add this to the storyboard.

File: modules/firebase/utils.ts (hunk 3 above)
Comment (arkadiuszbachorski) [reply modules/firebase/utils.ts:16]: Definitely, for now I just copy pasted it as quick solution.

File: packages/design-system/src/components/CopyText/CopyText.tsx
Comment (arkadiuszbachorski) [reply packages/design-system/src/components/CopyText/CopyText.tsx:1]: Every component is going to have story and tests, it's WIP! :)

**Comment by arkadiuszbachorski** [comment]
> I had a very quick first look! It would be great to get the new components into the storyboard to inspect more closely but everything already looks cool on the video you shared on Slack!

Everything will have their Storybooks, WIP!

> Could be that this is an issue with the backend authentication, I at least used a user with admin permission according to the Firebase datastore. Let me know what setup you use at the moment to render the components and to inspect the setup.

This was caused by wrong permission rules, the thing I raised on Slack. I just rollbacked to previous rules, it should work correctly now.

> We should setup a Firebase-Emulator enabled testing environment to make sure that we can easily test the mobile apps & web dashboard in a reproducible manner. CC @pauljohanneskraft.

Agree, now I am just using live Firebase data.

File: modules/firebase/utils.ts (hunk 3 above)
Comment (arkadiuszbachorski) [reply modules/firebase/utils.ts:16]: https://github.com/StanfordBDHG/ENGAGE-HF-Web-Frontend/issues/15

File: package.json (hunk 1 above)
Comment (arkadiuszbachorski) [reply package.json:37]: `firebase-admin` is used for retrieving user's auth data. We can't get it through Firestore. For example, to retrieve patients data, we need to combine data from user's auth and `user` collection. 

User can access other users data only if authenticated, for their respective roles and `firebase-admin` is executed on the server only. I agree it probably would be better if it was separate Firebase function, but that shouldn't be blocking for now.

File: modules/firebase/utils.ts (hunk 3 above)
Comment (PSchmiedmayer) [reply modules/firebase/utils.ts:16]: Thank you!

File: packages/design-system/src/components/CopyText/CopyText.tsx
Comment (PSchmiedmayer) [reply packages/design-system/src/components/CopyText/CopyText.tsx:1]: Thank you! 🚀

File: package.json (hunk 1 above)
Comment (PSchmiedmayer) [reply package.json:37]: @pauljohanneskraft Would it make sense & doable to extract that functionality in a cloud function so we can avoid that work in the frontend codebase & ensure that we don't need to embed admin SDK keys?

On a general level: I think we should generally use the firebase queries using the client SDK when possible to ensure that we build on top of the Security Rules but in some edge cases we should selectively use cloud functions if needed.

**Comment by PSchmiedmayer** [comment]
Thank you for continuing to improve this @arkadiuszbachorski! Sounds good; let's sync up with @pauljohanneskraft so we can combine this with the firebase setup to have a nice emulator testing environment.

File: packages/design-system/src/components/CopyText/CopyText.tsx
Comment (arkadiuszbachorski) [reply packages/design-system/src/components/CopyText/CopyText.tsx:1]: @PSchmiedmayer Done ✅

**Comment by arkadiuszbachorski** [comment]
@PSchmiedmayer I updated this PR with:

* tests and storybooks
* emulators support
* removed usage of `firebase-admin` in favor of calling `getUsersInformation` function
* resolved couple other todos, improvements

There is drop in coverage, because I didn't cover users and patients listing flow. It will be a separate E2E effort.

**Pull request merged**